blk/components/virt.c

```diff
@@ -184,201 +160,181 @@ static bool handle_mbr_reply()
-                err = blk_enqueue_resp(&h, BLK_RESP_OK, drv_success_count, cli_data.cli_req_id);
-                assert(!err);
-                break;
-            case BLK_REQ_WRITE:
-                err = blk_enqueue_resp(&h, BLK_RESP_OK, drv_success_count, cli_data.cli_req_id);
-                assert(!err);
-                break;
-            case BLK_REQ_FLUSH:
-            case BLK_REQ_BARRIER:
-                err = blk_enqueue_resp(&h, BLK_RESP_OK, drv_success_count, cli_data.cli_req_id);
-                assert(!err);
-                break;
+                    * include a corresponding libmicrokit API. */
+                seL4_ARM_VSpace_Invalidate_Data(3, reqbk.vaddr, reqbk.vaddr + (BLK_TRANSFER_SIZE * reqbk.count));
             }
-        } else {
-            // When more error conditions are added, this will need to be updated to a switch statement
-            err = blk_enqueue_resp(&h, drv_status, drv_success_count, cli_data.cli_req_id);
-            assert(!err);
+            break;
+        case BLK_REQ_WRITE:
+        case BLK_REQ_FLUSH:
+        case BLK_REQ_BARRIER:
+            break;
+        default:
+            /* This should never happen as we will have sanitized request codes before they are bookkept */
+            LOG_BLK_VIRT_ERR("bookkept client %d request code %d is invalid, this should never happen\n", reqbk.cli_id, reqbk.code);
+            assert(false);
         }
 
-        // Notify corresponding client
-        microkit_notify(clients[cli_data.cli_id].ch);
+        err = blk_enqueue_resp(&h, drv_status, drv_success_count, reqbk.cli_req_id);
+        assert(!err);
+
+        /* Notify corresponding client */
+        microkit_notify(clients[reqbk.cli_id].ch);
     }
 }
 
 static void handle_client(int cli_id)
 {
     blk_queue_handle_t h = clients[cli_id].queue_h;
-    uintptr_t cli_data_base = blk_virt_cli_data_region(blk_client_data_start, cli_id);
+    uintptr_t cli_data_base = blk_virt_cli_data_region(blk_client_data, cli_id);
     uint64_t cli_data_region_size = blk_virt_cli_data_region_size(cli_id);
 
-    blk_req_code_t cli_code;
-    uintptr_t cli_offset;
-    uint32_t cli_block_number;
-    uint16_t cli_count;
-    uint32_t cli_req_id;
-
-    uintptr_t drv_addr;
-    uint32_t drv_block_number;
-    uint32_t drv_req_id = 0;
+    blk_req_code_t cli_code = 0;
+    uintptr_t cli_offset = 0;
+    uint32_t cli_block_number = 0;
+    uint16_t cli_count = 0;
+    uint32_t cli_req_id = 0;
 
     int err = 0;
     while (!blk_queue_empty_req(&h)) {
         err = blk_dequeue_req(&h, &cli_code, &cli_offset, &cli_block_number, &cli_count, &cli_req_id);
         assert(!err);
 
+        blk_resp_status_t resp_status = BLK_RESP_ERR_UNSPEC;
+
+        if (ialloc_full(&ialloc)) {
+            LOG_BLK_VIRT_ERR("out of request indices to allocate\n");
+            resp_status = BLK_RESP_ERR_UNSPEC;
+            goto req_fail;
+        }
```

Comment:
Should this ever happen? Shouldn't we have an invariant that the driver queue is at least the size of all the client queues combined to not have to deal with conditions like this?

Reply:
We can have a situation where the driver dequeues all entries in a full req queue, and before enqueuing a response (say it was waiting for an i/o interrupt), the client sends more requests and notifies the virtualiser to forward these requests to the driver. In that scenario ialloc will be full and the queue is not. There is no guarantee how many requests a driver will process before it responds to the virtualiser.

Reply:
In the networking sub-system, this scenario is not possible since the capacity of the virt <-> driver queues is always at least the size of all the client queues. I think we should do the same here so you don't run into this case.

Reply:
I don't think setting the driver queue capacity to be more than the size of all client queues would avoid this problem. It's a problem with how many requests the driver decides to process prior to responding. The virtualiser can't free an id until its responded to by the driver, so we can have scenarios where the queues are empty (because the driver has dequeued them), and still receive a bunch more requests from the client, which will be more than the capacity of both client queues combined. We can't say how many requests the driver can dequeue before it responds, so we must always have a check for overflowing index allocations.

Reply:
Its about how many inflight requests there are, which we can't guarantee to be lower than a certain number*

Reply:
Hmm, I'm a bit confused.

If the client has N IDs and enqueues N buffers into the virtualiser, that means it has 0 free buffers left so cannot enqueue any more requests. The virtualiser does not give any free buffers back to the client until the driver gives the free buffers back to the virtualiser as well right?

Reply:
It isn't quite like the network queues where we have a free and active queue. Clients are free to make as many requests as it wants to the virtualiser. It can make requests as long as there is a free entry in the queue. Say the client enqueues N requests, whenever the virtualiser dequeues from the req queue the client is again free to enqueue mroe. Each request is not associated with a same size buffer region like the network queues, that's why the client must manage the memory themselves.

Reply:
Yes you're right.

What I'm wondering though if we should have the policy of dequeueing a request from the client if we cannot pass it to the driver.

Reply:
As a client this would confuse me, as whether or not I can enqueue a request or not depends on the behaviour of other clients.

The current policy would lead to correct clients having their requests returned with an error because the system is under load. That doesn't quite make sense to me.

Reply:
Agreed, it doesn't make sense to have system load result in errors for client requests. Thus we now make sure that in the virtualiser, ialloc full and driver req queue full conditions does not return error, instead we wait for notifications from driver to indicate free queue entries and free indices, and then we re-handle forwarding of client requests.